File: exercises/D107/solution.pas
var
    n, a, b: int64;
    square, triangular: boolean;
begin
    read(n);
    a := round(sqrt(n));
    b := trunc(sqrt(n * 2));
    square := a * a = n;
    triangular := b * (b + 1) div 2 = n;
    if (square and triangular) then
        writeln('Both')
    else if (square) then
        writeln('Square')
    else if (triangular) then
        writeln('Triangular')
    else
        writeln('Neither');
end.

File: exercises/D107/solution.py
issquare = lambda x: round(x**0.5)**2 == x
n = int(input())
square = issquare(n)
triangular = issquare(8*n+1)
# N is triangular if 8n+1 is a square number

if square and triangular:
    print("Both")
elif square:
    print("Square")
elif triangular:
    print("Triangular")
else:
    print("Neither")
